pydm/data_plugins/calc_plugin.py

```diff
@@ -0,0 +1,204 @@
+        ch_index : int
+            The channel index on the list for this rule.
+        value : bool
+            Whether or not this channel is connected.
+
+        """
+        self._connections[ch_index] = value
+        self._send_update(self.connected, self._value)
+
+    def warn_unconnected_channels(self):
+        logger.debug(
+            "Calculation '%s': Not all channels are connected, skipping execution.",
+            self.objectName())
+
+    def calculate_expression(self):
+        """
+        Evaluate the expression defined by the rule and emit the `rule_signal`
+        with the new value.
+        """
+        vals = list(self._values)
+
+        if None in vals:
+            logger.debug('Skipping execution as not all values are set.')
+            return
+
+        eval_env = {'math': math,
+                    'np': np,
+                    'ch': vals}
+        eval_env.update({k: v
+                         for k, v in math.__dict__.items()
+                         if k[0] != '_'})
+
+        try:
+            ret = eval(self._expression, eval_env)
+            self._value = ret
+            self._send_update(self.connected, ret)
+        except Exception as e:
+            logger.exception("Error while evaluating CalcPlugin connection %s",
+                             self.objectName())
+
+
+class Connection(PyDMConnection):
+    def __init__(self, channel, address, protocol=None, parent=None):
+        super(Connection, self).__init__(channel, address, protocol, parent)
+        self._calc_thread = None
+        self._configuration = {}
+        self._waiting_config = True
+
+        self.add_listener(channel)
+        self._init_connection()
+
+        self._setup_calc(address)
+
+    def _init_connection(self):
+        self.write_access_signal.emit(False)
+
+    def _setup_calc(self, address):
+        if not self._waiting_config:
+            logger.debug('CalcPlugin connection already configured.')
+            return
+        try:
+            self._configuration = json.loads(address)
+        except:
+            logger.info("Invalid configuration for CalcPlugin connection. %s",
+                        address)
+            return
+
+        if self._configuration.get('channels') \
+                and self._configuration.get('expr'):
+            self._waiting_config = False
+
+        name = self._configuration.get('name')
+
+        self._calc_thread = CalcThread(self._configuration)
+        self._calc_thread.setObjectName("calc_{}".format(name))
+        self._calc_thread.new_data_signal.connect(self.receive_new_data,
+                                                  Qt.QueuedConnection)
+        self._calc_thread.start()
+
+    @Slot(dict)
+    def receive_new_data(self, data):
+        if not data:
+            return
+        conn = data.get('connection', False)
+        val = data.get('value', None)
+        self.connected = conn
+        self.connection_state_signal.emit(conn)
```

Comment:
Always emitting this? Or should it be when `data.get('connection') is not None`?

Reply:
Addressed via https://github.com/slaclab/pydm/pull/637/commits/1ed6327c9ffcc39851e63cc83becb2d8a113914d